core/src/test/java/org/infinispan/tx/locking/RollbackDuringLockAcquisitionTest.java

```diff
@@ -0,0 +1,188 @@
+            {true, new JBossStandaloneJTAManagerLookup()},
+            {false, new EmbeddedTransactionManagerLookup()},
+            {false, new JBossStandaloneJTAManagerLookup()}
+      };
+   }
+
+   /*
+    * The "real world" conditions are:
+    *  - Narayana timeout <= lock timeout
+    *
+    * Cause/Steps:
+    *  - Narayana aborts a transaction which is waiting for a lock "L"
+    *  - The RollbackCommand is sent.
+    *    - Lock "L" is not locked by the transaction, so unlockAll(context.getLockedKeys(), context.getLockOwner());
+    *      never changes the state from WAITING => RELEASED
+    *  - When the lock "L" is released, the state changes from WAITING => ACQUIRED
+    *  - No RollbackCommand/CommitCommand is followed leaving the lock "L" in ACQUIRED state forever
+    */
+   @Test(dataProvider = "tm-lookup")
+   public void testRollbackWhileWaitingForLock(boolean useSynchronization, TransactionManagerLookup lookup)
+         throws Exception {
+      final String cacheName = concat("local-put", lookup.getClass().getSimpleName(), useSynchronization);
+      final String key = concat("reaper-put", lookup.getClass().getSimpleName(), useSynchronization);
+
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      builder.transaction()
+             .transactionManagerLookup(lookup)
+             .transactionMode(TransactionMode.TRANSACTIONAL)
+             .lockingMode(LockingMode.PESSIMISTIC)
+             .useSynchronization(useSynchronization)
+             .recovery().disable();
+      builder.locking()
+             .lockAcquisitionTimeout(TimeUnit.MINUTES.toMillis(1));
+
+      cacheManager.defineConfiguration(cacheName, builder.build());
+      Cache<String, String> cache = cacheManager.getCache(cacheName);
+      LockManager lockManager = extractLockManager(cache);
+      TransactionManager tm = cache.getAdvancedCache().getTransactionManager();
+
+      // simulates lock acquired by other transaction
+      lockManager.lock(key, "_tx_", 1, TimeUnit.SECONDS).lock();
+      assertLocked(cache, key);
+
+      tm.begin();
+      CompletableFuture<?> put = cache.putAsync(key, "value1");
+
+      GlobalTransaction gtx = extractComponent(cache, TransactionTable.class).getGlobalTransaction(tm.getTransaction());
+
+      // wait until the tx is queued in the InfinispanLock
+      eventually(() -> lockManager.getLock(key).containsLockOwner(gtx));
+
+      //abort the transaction, simulates the TM's reaper aborting a long running transaction
+      tm.rollback();
+
+      // release the lock. the "put" must not acquire the lock
+      lockManager.unlock(key, "_tx_");
+
+      Exceptions.expectCompletionException(InvalidTransactionException.class, put);
+
+      assertNotLocked(cache, key);
+      assertNoTransactions(cache);
+   }
+
+   @Test(dataProvider = "tm-lookup")
+   public void testRollbackWhileWaitingForLock2(boolean useSynchronization, TransactionManagerLookup lookup)
```

Comment:
Please add a description here, too, and/or rename the methods to something like `testRollbackWhileWaitingForLockDuringPut` and `testRollbackWhileWaitingForLockDuringLock`